src/docfx/build/toc/TocMap.cs

```diff
@@ -98,57 +98,76 @@ public IEnumerable<FilePath> GetFiles()
-        T file, IEnumerable<T> tocs, Dictionary<T, T[]> documentsToTocs, Func<T, string> getPath) where T : class, IComparable<T>
+        T file, T[] tocs, Dictionary<T, T[]> documentsToTocs, Func<T, string> getPath) where T : class, IComparable<T>
     {
-        var hasReferencedTocs = false;
+        bool hasReferencedTocs;
+
         var filteredTocs = (hasReferencedTocs = documentsToTocs.TryGetValue(file, out var referencedTocFiles)) ? referencedTocFiles : tocs;
+        if (filteredTocs is null || filteredTocs.Length <= 0)
+        {
+            return (default, false);
+        }
 
-        var tocCandidates = from toc in filteredTocs
-                            let dirInfo = GetRelativeDirectoryInfo(getPath(file), getPath(toc))
-                            where hasReferencedTocs || dirInfo.subDirectoryCount == 0 /*due breadcrumb toc*/
-                            select (subCount: dirInfo.subDirectoryCount, parentCount: dirInfo.parentDirectoryCount, toc);
+        var minCandidate = default((int subDirectoryCount, int parentDirectoryCount, T toc)?);
 
-        return (tocCandidates.DefaultIfEmpty().Aggregate((minCandidate, nextCandidate) =>
+        foreach (var toc in filteredTocs)
         {
-            var result = minCandidate.subCount - nextCandidate.subCount;
-            if (result == 0)
-            {
-                result = minCandidate.parentCount - nextCandidate.parentCount;
-            }
-            if (result == 0)
+            var (subDirectoryCount, parentDirectoryCount) = GetRelativeDirectoryInfo(getPath(file), getPath(toc));
+
+            // Due to breadcrumb toc
+            if (hasReferencedTocs || subDirectoryCount == 0)
             {
-                result = minCandidate.toc.CompareTo(nextCandidate.toc);
+                var candidate = (subDirectoryCount, parentDirectoryCount, toc);
+                if (minCandidate == null || candidate.CompareTo(minCandidate.Value) < 0)
+                {
+                    minCandidate = candidate;
+                }
             }
-            return result <= 0 ? minCandidate : nextCandidate;
-        }).toc, hasReferencedTocs);
+        }
+
+        return (minCandidate?.toc, hasReferencedTocs);
     }
 
     private static (int subDirectoryCount, int parentDirectoryCount) GetRelativeDirectoryInfo(string pathA, string pathB)
     {
-        var relativePath = PathUtility.NormalizeFile(Path.GetDirectoryName(PathUtility.GetRelativePathToFile(pathA, pathB)) ?? "");
-        if (string.IsNullOrEmpty(relativePath))
+        // Find common directory prefix
+        var commonStartIndex = 0;
+        var minLength = Math.Min(pathA.Length, pathB.Length);
+
+        for (var i = 0; i < minLength; i++)
         {
-            return default;
+            var chA = pathA[i];
+            var chB = pathB[i];
+            if (chA != chB)
+            {
+                break;
+            }
+
+            if (chA == '/')
```

Comment:
Assuming `pathA` is the shorter one here?

Reply:
No, `chA == chB` when it hit here